extension/ts/main/extensions/tags/validation.ts

```diff
@@ -0,0 +1,141 @@
+import {getAllTags, getAncestry} from "../../adapters/tags";
+import {createModal} from "../../ui/modal";
+import {loaderOverlaid} from "../../ui/loadingIndicator";
+import {isAuthorized} from "../author/util/isAuthorized";
+import {UIColour} from "../../ui/colour";
+import {OnSave} from "../../entities/bookForm";
+
+declare const SPREADSHEET_ID: string; // set by webpack
+
+type GetTagsOptions = {noCache: boolean};
+
+const createHighlighterComponents = () => {
+	const highlighter = document.createElement("div");
+	highlighter.className = "bookEditInput"; // this comes from LibraryThing. Stealing their css
+	highlighter.id = "vbl-text-highlighter";
+
+	const backdrop = document.createElement("div");
+	backdrop.id = "vbl-text-backdrop";
+	backdrop.append(highlighter);
+
+	return {highlighter, backdrop};
+};
+
+const applyHighlights = async (text: string): Promise<string> => {
+	const validTags = await getAllTags();
+	const futureParts = text.split(",").map(async (part) => {
+		const trimmedPart = part.trim();
+		if (validTags.has(trimmedPart)) {
+			return part;
+		} else {
+			const mark = document.createElement("mark");
+			mark.innerText = trimmedPart;
+			return part.replace(trimmedPart, mark.outerHTML);
+		}
+	});
+	const parts = await Promise.all(futureParts);
+	return parts.join(",") + "\n";
+};
+
+const handleInput = (tagInput: HTMLTextAreaElement, highlighter: HTMLElement) => {
+	const handler = async () => (highlighter.innerHTML = await applyHighlights(tagInput.value));
+	tagInput.addEventListener("input", handler);
+	tagInput.addEventListener("change", handler);
+	return handler();
+};
+
+const handleViewChange = (tagInput: HTMLTextAreaElement, backdrop: HTMLElement) => {
+	const handler = () => {
+		backdrop.scrollTop = tagInput.scrollTop;
+		backdrop.style.width = tagInput.clientWidth + "px";
+		backdrop.style.height = tagInput.clientHeight + "px";
+	};
+	tagInput.addEventListener("scroll", handler);
+	tagInput.addEventListener("resize", handler);
+	return handler();
+};
+
+const getInvalidTags = async (tags: string[], options: GetTagsOptions): Promise<string[]> => {
+	const validTags = await getAllTags(options);
+	return tags.filter((tag) => !validTags.has(tag));
```

Comment:
do we want this to be case insensitive or should it match exactly what's in the tag index?

Reply:
but then we start getting this from libarything :(
<img width="642" alt="image" src="https://user-images.githubusercontent.com/35436247/200194297-affb0940-e4d2-4995-b1a2-5a868022095a.png">

however those two tags are considered the "same" in the librarything search engine.

Reply:
I think for now, we just leave it as is, and then in the future we automatically fix people's bad casing